index.rst

```diff
@@ -13,17 +13,542 @@
+
+Workspaces are expected to play a central role in the DMTN-249 :cite:`DMTN-249` consistency model, in which file artifacts are written prior to the insertion of their metadata into the database, but only after some record is made in the central repository that those artifacts *may* exist.
+This means workspace construction and removal operations need to be very careful about consistency in the presence of errors and concurrency.
+This is complicated by the fact that workspaces are extensible; all concrete implementations will live downstream of ``daf_butler``.
+
+We envision workspace creation to be aided by two helper classes that are typically defined along with a new :class:`WorkspaceButler` implementation:
+
+- A :class:`WorkspaceFactory` is a callable that creates a new workspace from a full butler and some standard user-provided arguments, returning a :class:`WorkspaceButler` instance and a :class:`WorkspaceConfig` instance.
+- :class:`WorkspaceConfig` is a pydantic model that records all state needed to make a new :class:`WorkspaceButler` client to an existing workspace, with a method to create a :class:`WorkspaceButler` from that state.
+
+After a factory is used to write the rest of the workspace's initial state, the configuration is written to a JSON file in a predefined (by a URI template in the central repository butler configuration) location.
+This file is read and used to make a new :class:`WorkspaceButler` instance without requiring access to the full butler.
+
+This is all driven by a :meth:`~Butler.make_workspace` method on the full :class:`~lsst.daf.butler.Butler` class, for which the prototyping here includes a nominal implementation with detailed comments about how responsibility (especially for error-handling) is shared by different methods.
+
+After an internal workspace has been created, a client :class:`WorkspaceButler` can be obtained from a full :class:`Butler` to the central repository by calling :meth:`Butler.get_workspace`, or without ever making a full butler by calling :meth:`ButlerConfig.make_workspace_butler`.
+External workspace construction goes through :meth:`WorkspaceButler.make_external`.
+
+We expect most concrete workspace implementations to define a ``butler`` subcommand for their creation, and for most end users to interact only with that command-line interface.
+
+Committing a workspace goes through :meth:`WorkspaceButler.commit`, and abandoning one goes through :meth:`WorkspaceButler.abandon`.
+Creating a new standalone repository goes through :meth:`WorkspaceButler.export`.
+All of these have nominal implementations in the prototype, showing that they delegate to many of the same abstract methods to transfer their content to full butlers and remove anything that remains from the workspace.
+
+To make exporting and external workspaces more useful, :meth:`WorkspaceButler.transfer_inputs` may be used to transfer the file artifacts of input datasets used by the workspace into the workspace itself.
```

Comment:
I guess this will also change Datastore configuration for workspace butler from whatever it was in central repo to local Posix Datastore?

Reply:
Yes, that makes sense.